### PR TITLE
feat: oauth provider

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

(A few sentences about this PR)

## Related issues

-   Link to issue1 here
-   Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
-   [ ] Make sure that `syncio` / `asyncio` functions are consistent.
-   [ ] If access token structure has changed
    -   Modified test in `tests/sessions/test_access_token_version.py` to account for any new claims that are optional or omitted by the core

## Remaining TODOs for this PR

-   [ ] Item1
-   [ ] Item2

### CODE REVIEW

File: supertokens_python/recipe/oauth2provider/recipe_implementation.py [hunk 1]
```diff
@@ -0,0 +1,1035 @@
...
+        token: str,
+        requirements: Optional[OAuth2TokenValidationRequirements],
+        check_database: Optional[bool],
+        user_context: Dict[str, Any],
+    ) -> Dict[str, Any]:
+        access_token_obj = parse_jwt_without_signature_verification(token)
+
+        # Verify token signature using session recipe's JWKS
+        session_recipe = SessionRecipe.get_instance()
+        matching_keys = get_latest_keys(session_recipe.config, access_token_obj.kid)
```
Comment (porcellus) [inline supertokens_python/recipe/oauth2provider/recipe_implementation.py:642]: this goes through caching right?

**Review by porcellus — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/oauth2provider/recipe_implementation.py (hunk 1 above)
Comment (sattvikc) [reply supertokens_python/recipe/oauth2provider/recipe_implementation.py:642]: yes it does

File: supertokens_python/framework/django/django_response.py [hunk 2]
```diff
@@ -88,3 +88,9 @@ def set_json_content(self, content: Dict[str, Any]):
                 separators=(",", ":"),
             ).encode("utf-8")
             self.response_sent = True
+
+    def redirect(self, url: str) -> BaseResponse:
+        if not self.response_sent:
+            self.set_header("Location", url)
+            self.set_status_code(302)
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply supertokens_python/framework/django/django_response.py:95]: yes

File: supertokens_python/framework/fastapi/fastapi_response.py [hunk 3]
```diff
@@ -94,3 +94,9 @@ def set_json_content(self, content: Dict[str, Any]):
             self.set_header("Content-Length", str(len(body)))
             self.response.body = body
             self.response_sent = True
+
+    def redirect(self, url: str) -> BaseResponse:
+        if not self.response_sent:
+            self.set_header("Location", url)
+            self.set_status_code(302)
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply supertokens_python/framework/fastapi/fastapi_response.py:101]: yes

**Pull request merged**